### PR TITLE
Allow union-with-callable attributes to be overridden by methods

### LINKED ISSUE #12569: mypy incorrectly marks a @staticmethod definition of an attribute A to be incompatible with A's signature in a parent class.
**Bug Report**
When a class **C** defines an attribute **a** to be a union type with `Callable` as one of its types (`a: Callable | ...`), **mypy** incorrectly marks a `@staticmethod` definition of **a** in a child class that inherits from **C** to be incompatible with **a**'s signature in **C**.

Worth noting that the following code produces no errors:
```python
from typing import Callable


class Parent:
    foo: Callable[[str], str]

class Child2(Parent):
    @staticmethod
    def foo(x: str) -> str:
        return x.lower()
```

but the following code does

```python
from typing import Callable


class Parent:
    foo: Callable[[str], str] | str 


class Child1(Parent):
    foo = "bar"  # OK


class Child2(Parent):
    @staticmethod
    def foo(x: str) -> str:  # Signature of "foo" incompatible with supertype "Parent"
        return x.lower()
```

**Expected Behavior**
No errors.

**Actual Behavior**
```shell
example.py: note: In class "Child2":
example.py:14: error: Signature of "foo" incompatible with supertype "Parent"  [override]
Found 1 error in 1 file (checked 1 source file)
```
**Environment**

- Mypy version used: `0.941`
- Mypy command-line flags: `--strict`, `--show-error-codes`,  `--show-error-context`
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: `3.10.3`
- Operating system and version: MacOS 12.3

### PR DESCRIPTION
**Pull request by brianschubert**
Fixes #12569

### Before
```python
from typing import Callable

class Parent:
    f1: Callable[[str], str]
    f2: Callable[[str], str] | str

class Child(Parent):
   def f1(self, x: str) -> str: return x  # ok
   def f2(self, x: str) -> str: return x  # Signature of "f2" incompatible with supertype "Parent"
```

### After
```python
from typing import Callable

class Parent:
    f1: Callable[[str], str]
    f2: Callable[[str], str] | str

class Child(Parent):
   def f1(self, x: str) -> str: return x  # ok
   def f2(self, x: str) -> str: return x  # ok
```

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -2154,6 +2154,11 @@ def check_method_override_for_base_with_name(
                     override_class_or_static,
                     context,
                 )
+            elif isinstance(original_type, UnionType) and any(
```
Comment (JelleZijlstra) [inline mypy/checker.py:2157]:
```suggestion
            elif isinstance(original_type,UnionType) and any(
```
Testing if pre-commit.ci works now

**Review by hauntsaninja — COMMENTED** [review]
Nice! Could you add a test case with this error code enabled: https://github.com/python/mypy/pull/16399/files

**Review by hauntsaninja — APPROVED** [review]
Thanks!

I kind of want to try `ignore_pos_arg_names=False`. This is something we probably want to eventually do anyway and should be fairly harmless in this case. But it's likely wiser to be consistent.

**Comment by hauntsaninja** [comment]
I was thinking about this PR some more and I think it introduces some unsoundness. We should have a Liskov error in the following:
```
import typing

class X:
    @property
    def x(self) -> typing.Optional[bool]: ...

    @x.setter
    def x(self, value: typing.Optional[bool]) -> None: ...

class Y(X):
    @property
    def x(self) -> bool: ...

    @x.setter
    def x(self, value: bool) -> None: ...
```
Input to setter should be wider, output of getter should be narrower.

**Comment by brianschubert** [comment]
Hmm, I think that setup is currently caught by mutable-override, is it not? Is the behavior after this PR different from what mypy does for this setup?
```python
class X:
    x: int | None

class Y(X):
    x: int
```

**Comment by hauntsaninja** [comment]
Oh looks like I must have messed up my test. It does error with mutable-override, sorry for the noise!

**Comment by hauntsaninja** [comment]
I think what happened is I messed up the branches I tested. I think I probably did this, which mypy doesn't error on:
```
class AA:
    @property
    def a(self) -> int: ...
    @a.setter
    def a(self, value: int | str) -> None: ...

class BB(AA):
    @property
    def a(self) -> int: ...
    @a.setter
    def a(self, value: int) -> None: ...
```
This would be unsound, except that mypy doesn't support this at all and will error if you attempt to set `a` to str. This is the (very popular) issue for that: https://github.com/python/mypy/issues/3004

**Pull request merged**